src/assert_not_impl.rs

```diff
@@ -0,0 +1,115 @@
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {}
+/// assert_not_impl!(str; *const u8, Sync, From<&'static str>);
+/// assert_not_impl!(vec; *const u8, Sync);
+/// ```
+///
+/// The [labeling limitation](index.html#limitations) is not necessary if
+/// compiling on nightly Rust with the `nightly` feature enabled:
+///
+#[cfg_attr(feature = "nightly", doc = "```")]
+#[cfg_attr(not(feature = "nightly"), doc = "```ignore")]
+/// #![feature(underscore_const_names)]
+/// # #[macro_use] extern crate static_assertions;
+///
+/// assert_not_impl!(&'static mut u8, Copy);
+///
+/// fn main() {
+///     assert_not_impl!(u32, Into<usize>);
+/// }
+/// ```
+///
+/// The following example fails to compile since `u32` can be converted into `u64`.
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {
+/// assert_not_impl!(u32, Into<u64>);
+/// # }
+/// ```
+///
+/// Note that all traits given in a single invocation need to not be implemented. If you want to
+/// check that none of multiple trait are implemented you must invoke the macro several times.
+///
+#[cfg_attr(feature = "nightly", doc = "```ignore")]
+#[cfg_attr(not(feature = "nightly"), doc = "```")]
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // Cell<u32> is not both Sync and Send
+/// assert_not_impl!(cell; Cell<u32>, Sync, Send);
+/// ```
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // But it is Send, this fails to compile
+/// assert_not_impl!(cell; Cell<u32>, Send);
+/// ```
+///
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [`Sync`]: https://doc.rust-lang.org/std/marker/trait.Sync.html
+/// [blanket]: https://doc.rust-lang.org/book/second-edition/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods
+#[macro_export(local_inner_macros)]
+macro_rules! assert_not_impl {
+    ($($xs:tt)+) => { _assert_not_impl!($($xs)+); };
+}
+
+#[doc(hidden)]
+#[cfg(feature = "nightly")]
+#[macro_export(local_inner_macros)]
+macro_rules! _assert_not_impl {
+    ($x:ty, $($t:path),+ $(,)*) => {
+        const _: fn() -> () = || {
+            // Ensures there can not be any conflict.
+            struct Sealed;
+            struct Check<T: ?Sized>(T);
+
+            trait Fine { fn test_for_assert_not_impl_conflict(Sealed) { } }
+            trait Conflict { fn test_for_assert_not_impl_conflict(Sealed) { } }
+
+            impl<T: ?Sized> Fine for Check<T> { }
+            impl<T: ?Sized $(+ $t)*> Conflict for Check<T> { }
+
+            Check::<$x>::test_for_assert_not_impl_conflict(Sealed)
+        };
+    };
+}
+
+#[doc(hidden)]
+#[cfg(not(feature = "nightly"))]
+#[macro_export(local_inner_macros)]
+macro_rules! _assert_not_impl {
+    ($x:ty, $($t:path),+ $(,)*) => {
+        {
```

Comment:
Is this block necessary since this macro arm is already nested inside `fn $label() { ... }` from the other macro arm?

Reply:
I had simply taken the style from `assert_impl!` but it probably is not strictly necessary.

Reply:
I'm not familiar enough with the macro hygiene rules to know how this would be affected by not having a scope so I'd like to keep this scoped just for sanity.